pkg/cli/longhorn.go

```diff
@@ -0,0 +1,243 @@
+package cli
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+	"time"
+
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	lhv1b1 "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+)
+
+var (
+	scaleDownReplicasWaitTime = time.Minute
+)
+
+const (
+	longhornNamespace       = "longhorn-system"
+	overProvisioningSetting = "storage-over-provisioning-percentage"
+)
+
+func NewLonghornCmd(cli CLI) *cobra.Command {
+	return &cobra.Command{
+		Use:   "longhorn",
+		Short: "Perform operations on a longhorn installation within a kURL cluster",
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			return cli.GetViper().BindPFlags(cmd.PersistentFlags())
+		},
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return cli.GetViper().BindPFlags(cmd.Flags())
+		},
+	}
+}
+
+func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
+	return &cobra.Command{
+		Use:          "prepare-for-migration",
+		Short:        "Prepares Longhorn for migration to a different storage provisioner.",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			log.Print("Preparing Longhorn deployment for migration.")
+			cli, err := client.New(config.GetConfigOrDie(), client.Options{})
+			if err != nil {
+				return fmt.Errorf("error creating client: %s", err)
+			}
+			lhv1b1.AddToScheme(cli.Scheme())
+
+			var nodes corev1.NodeList
+			if err := cli.List(cmd.Context(), &nodes); err != nil {
+				return fmt.Errorf("error listing kubernetes nodes: %s", err)
+			} else if len(nodes.Items) == 1 {
+				log.Print("Only one node found, scaling down the number of volume replicas to 1.")
+				if err := scaleDownReplicas(cmd.Context(), cli); err != nil {
+					return fmt.Errorf("error scaling down longhorn replicas: %s", err)
+				}
+			}
+
+			unhealthy, err := unhealthyVolumes(cmd.Context(), cli)
```

Comment:
should we not to do this check before scale down?
So that we do not scale down, check that is not health and then stop to migration leaving the replicas scaled down?

Reply:
In a single node cluster the number of replicas is set to 3, that leaves all volumes in an unhealthy state as the replicas can't be scheduled. If we don't scale them down we can't check if they are healthy.

Reply:
Can we revert that before stop or just make clear that we did that to allow users came back if they need to?

Reply:
That's a good idea. I added a command that allows users to rollback the number of replicas if necessary. My reasoning is that the script changes the number of replicas and may find that some volumes are not healthy. If we automatically scale them up, it can become complicated for the user to debug, as other volumes may become unhealthy, or the automatic scale-up behavior may change the reason why the original unhealthy volume was unhealthy. Would you mind taking a look at it ? Great catch this one.